src/stdlib_stats_median.fypp

```diff
@@ -0,0 +1,217 @@
+              #:for fj in range(fi+1, rank+1)
+                do j${"_" * fj}$ = 1, size(x, ${fj}$)
+              #:endfor
+                  x_tmp(:) = x${select_subvector(rank, fi)}$
+                  call sort(x_tmp)
+
+                  if (mod(n, 2) == 0) then
+                      res${reduce_subvector(rank, fi)}$ = &
+                      #:if t1[0] == 'r'
+                          sum(x_tmp(c:c+1)) / 2._${o1}$
+                      #:else
+                          sum(real(x_tmp(c:c+1), kind=${o1}$) ) / 2._${o1}$
+                      #:endif
+                  else
+                      res${reduce_subvector(rank, fi)}$ = x_tmp(c)
+                  end if
+            #:for fj in range(1, rank)
+              end do
+            #:endfor
+          #:endfor
+          case default
+            call error_stop("ERROR (median): wrong dimension")
+        end select
+
+      end function ${RName}$
+    #:endfor
+  #:endfor
+
+
+  #:for k1, t1, o1 in IR_KINDS_TYPES_OUTPUT
+    #:for rank in RANKS
+      #:set RName = rname('median_mask_all',rank, t1, k1, o1)
+      module function ${RName}$(x, mask) result(res)
+        ${t1}$, intent(in) :: x${ranksuffix(rank)}$
+        logical, intent(in) :: mask${ranksuffix(rank)}$
+        real(${o1}$) :: res
+
+        integer(kind = int64) :: c, n
+        ${t1}$, allocatable   :: x_tmp(:)
+
+        if (any(shape(x) .ne. shape(mask))) then
+            call error_stop("ERROR (median): shapes of x and mask are different")
+        end if
+
+        x_tmp = pack(x, mask)
+
+        call sort(x_tmp)
+
+        n = size(x_tmp, kind=int64 )
+        c = floor( (n + 1) / 2._${o1}$, kind=int64 )
+
+        if (n == 0) then
+            res = ieee_value(1._${o1}$, ieee_quiet_nan)
+        else if (mod(n, 2_int64) == 0) then
+          #:if t1[0] == 'r'
+            res = sum(x_tmp(c:c+1)) / 2._${o1}$
+          #:else
+            res = sum(real(x_tmp(c:c+1), kind=${o1}$)) / 2._${o1}$
+          #:endif
+        else if (mod(n, 2_int64) == 1) then
+            res = x_tmp(c)
+        end if
+
+      end function ${RName}$
+    #:endfor
+  #:endfor
+
+  #:for k1, t1, o1 in IR_KINDS_TYPES_OUTPUT
+    #:for rank in RANKS
+      #:set RName = rname('median_mask',rank, t1, k1, o1)
+      module function ${RName}$(x, dim, mask) result(res)
+        ${t1}$, intent(in) :: x${ranksuffix(rank)}$
+        integer, intent(in) :: dim
+        logical, intent(in) :: mask${ranksuffix(rank)}$
+        real(${o1}$) :: res${reduced_shape('x', rank, 'dim')}$
+
+        integer(kind = int64) :: c, n
+        #:if rank > 1
+        #:for fj in range(1, rank+1)
+        integer :: j${"_" * fj}$
+        #:endfor
+        #:endif
+        ${t1}$, allocatable :: x_tmp(:)
+
+        if (any(shape(x) .ne. shape(mask))) then
+            call error_stop("ERROR (median): shapes of x and mask are different")
+        end if
+
+        select case(dim)
+          #:for fi in range(1, rank+1)
+          case(${fi}$)
```

Comment:
As per the comment above, suggest to add a comment `! Loop over every dimension of the array except dim`, and also, consider if you can collapse the 2 fypp loops to 1.

Reply:
done